hub/config/src/main/java/uk/gov/ida/hub/config/data/S3ConfigSource.java

```diff
@@ -1,81 +1,122 @@
+            return EMPTY_COLLECTION;
         }
-        return new S3ConfigSource(configConfiguration, s3Client);
     }
 
-    public RemoteConfigCollection getRemoteConfig() {
-        S3Object fullObject = s3Client.getObject(
-                new GetObjectRequest(configConfiguration.getSelfService().getS3BucketName(),
-                configConfiguration.getSelfService().getS3ObjectKey()));
-        InputStream s3ObjectStream =  fullObject.getObjectContent();
-        ObjectMapper om = new ObjectMapper();
-        DateFormat df = new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss");
-        om.setDateFormat(df);
-        SimpleModule module = new SimpleModule();
-        module.addDeserializer(RemoteConfigCollection.class, new RemoteConfigCollectionDeserializer());
-        om.registerModule(module);
+    private RemoteConfigCollection mapToRemoteConfigCollection(InputStream inputStream, Date lastModified) throws IOException {
         try {
-            return om.readValue(s3ObjectStream, RemoteConfigCollection.class);
-        } catch(IOException e) {
-            LOG.error("An error occured trying to get or process object {} from S3 Bucket {}",
-                    configConfiguration.getSelfService().getS3ObjectKey(),
-                    configConfiguration.getSelfService().getS3BucketName(), e);
-            throw new RuntimeException(e);
+            SelfServiceMetadata metadata = objectMapper.readValue(inputStream, SelfServiceMetadata.class);
+            RemoteConfigCollection remoteConfig = new RemoteConfigCollection(lastModified, metadata);
+            return remoteConfig;
         } finally {
-            if(fullObject != null) {
-                try {
-                    fullObject.close();
-                } catch(IOException e) {
-                    throw new RuntimeException(e);
-                }
-            }
+            inputStream.close();
         }
     }
 
+    private S3Object getS3Object(String key) {
+        return s3Client.getObject(
+                new GetObjectRequest(selfServiceConfig.getS3BucketName(),
+                        key));
+    }
+
+    private class S3ConfigCacheLoader extends CacheLoader<String, RemoteConfigCollection>{
+
+        @Override
+        public RemoteConfigCollection load(String key) throws Exception {
+            S3Object s3Object = getS3Object(key);
+            try{
+                return mapToRemoteConfigCollection(s3Object.getObjectContent(), s3Object.getObjectMetadata().getLastModified());
+            } catch(IOException e) {
+                LOG.error("An error occurred trying to get or process object {} from S3 Bucket {}",
+                        selfServiceConfig.getS3ObjectKey(),
+                        selfServiceConfig.getS3BucketName(), e);
+                throw(e);
+            }
+        }
+
+        @Override
+        public ListenableFuture<RemoteConfigCollection> reload(String key, RemoteConfigCollection existingConfig) {
```

Comment:
I think you're making this asynchronous at the wrong point. You want the most expensive operation, the network request to S3, to be made asynchronous or you would not get any benefit out of overriding this method. The asynchronous task should start at the top of this method before the S3 call.

Also, given how much cheap it is to move small amount of data around in memory compared to a network request you might want to consider ignoring all the lastModified logic.

Reply:
OK, I've made the whole lot async. I think the lastModified is a useful optimisation.

Reply:
You need to include L100 in the Future as that's the most expensive action.

Reply:
OK, I'd miss-understood what was happening here.

Reply:
Done